### PR TITLE
#61 EC2 Tag Renaming Support

### PR DESCRIPTION
**Pull request by joshuaroot**
* Added action to allow for renaming of tags on EC2 instances

### CODE REVIEW

File: c7n/resources/ec2.py [hunk 1]
```diff
@@ -634,6 +634,69 @@ def process_volume_set(self, resource):
...
+        Move source tag value to destination tag value
+
+        - Collect value from old tag
+        - Delete old tag
+        - Create new tag & assign stored value
+        """
+
+        source_tag = self.data.get('source')
+        tag_map = {t['Key']: t for t in instance.get('Tags', [])}
+        source = tag_map[source_tag]
```
Comment (kapilt) [inline c7n/resources/ec2.py]: what if source doesn't exist, this throws a key error.

File: c7n/resources/ec2.py [hunk 2]
```diff
@@ -634,6 +634,69 @@ def process_volume_set(self, resource):
...
+                if t['Key'] == src:
+                    filtered.append(i)
+                    break
+        resources = filtered
+        self.log.info("Filtered from %d instance(s) to %d" % (
+            count, len(resources)))
+        self.log.info("Renaming %s to %s on %d instance(s)" % (
+            src, dest, len(filtered)))
+
+        client = utils.local_session(self.manager.session_factory).client('ec2')
```
Comment (kapilt) [inline c7n/resources/ec2.py]: this needs to be done in the worker function being executed by the thread factory.

File: c7n/resources/ec2.py [hunk 3]
```diff
@@ -634,6 +634,69 @@ def process_volume_set(self, resource):
...
+        src = self.data.get('source')
+        dest = self.data.get('dest')
+        count = len(resources)
+        filtered = []
+        for i in resources:
+            for t in i.get('Tags'):
+                if t['Key'] == src:
+                    filtered.append(i)
+                    break
+        resources = filtered
```
Comment (kapilt) [inline c7n/resources/ec2.py]: this filtering logic would be nice to move to a helper method.

**Comment by kapilt** [comment]
The big question to me on this branch, is manipulating multiple tag/values on a resource, ie i'd rather the user provide mapping of source / dest so we can cover multiple keys in a single go.

File: c7n/resources/ec2.py (hunk 3 above)
Comment (kapilt) [inline c7n/resources/ec2.py]: not a big deal but it feels wierd to process the list twice, we could inline this at the modification site and use the futures results to do the reporting on total filtering count.

**Comment by kapilt** [comment]
There's an opportunity to make this much more efficient by grouping the resources with similiar values, such that your making total count of api calls not based on number of resources, but instead by number of unique values.

**Comment by kapilt** [comment]
Given we have mostly generic tag ops for other resources with subclasses doing resource specific logic, its worth thinking how this would fit into a similiar model given many taggable resources.

File: c7n/resources/ec2.py [hunk 4]
```diff
@@ -634,6 +634,69 @@ def process_volume_set(self, resource):
...
+        - Delete old tag
+        - Create new tag & assign stored value
+        """
+
+        source_tag = self.data.get('source')
+        tag_map = {t['Key']: t for t in instance.get('Tags', [])}
+        source = tag_map[source_tag]
+        destination_tag = self.data.get('dest')
+
+        client.delete_tags(
```
Comment (kapilt) [inline c7n/resources/ec2.py]: of course to do that safely, you'll need to sanity check that the resource is under tag count limit.

**Comment by kapilt** [comment]
if we do the group by value for mods, i think we can omit the multi-tag capability.

File: c7n/resources/ec2.py (hunk 4 above)
Comment (joshuaroot) [inline c7n/resources/ec2.py]: I wasn't sure we needed to perform a tag check at this point with the increase in the number of tags, but I can add it.

File: c7n/resources/ec2.py (hunk 3 above)
Comment (joshuaroot) [inline c7n/resources/ec2.py]: Now that you said that, it makes sense.

File: tests/test_ec2.py [hunk 5]
```diff
@@ -310,6 +310,20 @@ def test_ec2_untag(self):
...
+        policy = self.load_policy({
+            'name': 'ec2-rename-tag',
+            'resource': 'ec2',
+            'actions': [{
+                'type': 'rename-tag',
+                'map': [{
+                    'Success': 'Testing'}]
+            }]}, session_factory=session_factory)
+        resources = policy.run()
+        self.assertEqual(len(resources), 3)
```
Comment (kapilt) [inline tests/test_ec2.py]: This is not a valid test, it hasn't verified anything. how does it assert that the tags have been renamed?

**Comment by kapilt** [comment]
schema validation failed

File: c7n/resources/ec2.py [hunk 6]
```diff
@@ -653,6 +653,82 @@ def process_volume_set(self, resource):
...
+        self.delete_tag(
+            c, [r['InstanceId'] for r in resources], source_tag, source_value)
+        
+        if over:
+            self.create_tag(c, over, dest_tag, source_value)
+
+    def create_map(self, resources):
+        tag_map = self.data.get('map', [{}])[0]
+        keys = {}
+        for t in tag_map.keys():
```
Comment (kapilt) [inline c7n/resources/ec2.py]: this logic feels convoluted. your iterating over one map to populate another, you'll never hit a case where its in the one being populated.

File: c7n/resources/ec2.py [hunk 7]
```diff
@@ -653,6 +653,82 @@ def process_volume_set(self, resource):
...
+            if t not in keys:
+                keys[t] = {}
+        for r in resources:
+            tags = {t['Key']: t for t in r.get('Tags', [])}
```
Comment (kapilt) [inline c7n/resources/ec2.py]: t['Value'] instead of t

File: c7n/resources/ec2.py [hunk 8]
```diff
@@ -653,6 +653,82 @@ def process_volume_set(self, resource):
...
+            for k in keys:
+                if k not in tags:
+                    continue
+                value = tags[k]['Value']
+                if value not in keys[k]:
```
Comment (kapilt) [inline c7n/resources/ec2.py]: keys[k].setdefault(value, []).append(r)

File: c7n/resources/ec2.py (hunk 6 above)
Comment (joshuaroot) [inline c7n/resources/ec2.py]: You are 100% correct.

**Pull request closed without merging**